File: 0701-0800/0738-Monotone Increasing Digits/0738-Monotone Increasing Digits.cpp
class Solution {
public:
    int monotoneIncreasingDigits(int N) {
        string s = to_string(N);
        int j = -1;
        for (int i = 0; i < s.size() - 1; ++i) {
            if (s[i] > s[i + 1]) {
                j = i;
                break;
            }
        }
        
        if (j == -1) {
            return N;
        }
        
        for (int k = j; k >= 1; --k) {
            if (s[k - 1] > s[k] - 1) {
                --j;
            }
            else {
                break;
            }
        }
        
        --s[j];
        for (int k = j + 1; k < s.size(); ++k) {
            s[k] = '9';
        }
        return stoi(s);
    }
};

File: 0701-0800/0738-Monotone Increasing Digits/0738-Monotone Increasing Digits.py
class Solution:
    def monotoneIncreasingDigits(self, N: int) -> int:
        s = list(str(N))
        i = 0
        while i < len(s) - 1:
            if s[i] > s[i + 1]:
                break
            i += 1
        if i == len(s) - 1:
            return N

        for k in range(i, 0, -1):
            if int(s[k - 1]) > int(s[k]) - 1:
                i -= 1
            else:
                break
        s[i] = str(int(s[i]) - 1)
        s[i+1:] = '9'*(len(s) - i - 1)
        return int(''.join(s))
